modules/post/windows/manage/execute_dotnet_assembly.rb

```diff
@@ -125,7 +138,30 @@ def run
     if check_requirements(rclr, installed_dotnet_versions) == false
       fail_with(Failure::BadConfig, 'CLR required for assembly not installed')
     end
-    execute_assembly(exe_path)
+
+    if sysinfo.nil?
+      fail_with(Failure::BadConfig, 'Session invalid')
+    else
+      print_status("Running module against #{sysinfo['Computer']}")
+    end
+
+    execute_assembly(exe_path, rclr)
+
+    # This has to be here, not in cleanup, because if an exception is raised
+    # on the MSF side, we don't want to prematurely clean up the DLL that may
+    # still be running.
```

Comment:
If you update the C++ code to exit by calling [`ReflectiveFreeAndExitThread`](https://github.com/rapid7/metasploit-payloads/blob/master/c/meterpreter/source/dump_sam/ReflectiveFreeAndExitThread.c#L12) I think you can not only avoid this issue but also free the memory that the loader itself allocates [here.](https://github.com/rapid7/ReflectiveDLLInjection/blob/master/dll/src/ReflectiveLoader.c#L312)

When I added that code in rapid7/metasploit-payloads#631, I tested it pretty thoroughly on a few different versions of Windows including x86 and x64.

Reply:
Do you mean to copy that function into the HostingCLR solution and call it from there?

Reply:
Yes. Until we move it into a central location in the ReflectiveDLLInjection repository where it probably belongs, we'll need to copy it around. AFAIK, this will be the second time it's used so the first time it's been copied.

Reply:
I've done that, and am now freeing the three chunks of memory (assembly+params, reflectively loaded DLL and bootstrapping code) in the C++ side.